pennylane_lightning_gpu/src/algorithms/AdjointDiffGPU.hpp

```diff
@@ -217,19 +235,15 @@ template <class T = double> class AdjointJacobianGPU {
      */
     inline void updateJacobian(const StateVectorCudaManaged<T> &sv1,
                                const StateVectorCudaManaged<T> &sv2,
-                               std::vector<std::vector<T>> &jac,
-                               T scaling_coeff, size_t obs_index,
-                               size_t param_index) {
+                               decltype(cuUtil::getCudaType(T{})) *jac) {
```

Comment:
Ah, your goal is to replicate this entirely on the device, I missed that. In this instance, since this is private, I am fine with the pointer. Though, if it is possible to use a `DataBuffer` instead, that would be preferable. If not, feel free to keep as is.

I am fine with pointer arithmetic in this case too, as long as the intention is clear. Though, if preferable, can this be offloaded to `std::advance` to perform the stepping, as this improves clarity and intention.

Reply:
`std::advance` is applicable, so we can change it. However, I find `std::advance` more obfuscating than a simple `typed-ptr + offset`.

I see the main advantage of `std::advance` in it's "genericity": it works on all iterators (including non-random-access iterators) and pointers. This can be very handy when writing generic algorithms.

If you prefer `std::advance` nevertheless, please let us know, then we'll use `std::advance` it.

Reply:
Sure, let's keep it as pointer + offsets. :+1: 

Usually my concern with this is many non C++ folks tend to explore the codebase, and handing responsibility of pointer works to STL methods helps with them finding documentation of what/why something is happening. But, long-story shortened, fine to use typed-ptr + offset in this instance. We can always update in future if it makes sense to do so. Thanks!